infrastructure/infrastructure/utilities/git.py

```
@@ -190,302 +214,372 @@ def setRemoteURL(remote, url, path, logger):
                           logger=logger)
 
 
-def clone(gitURL, **kwargs):
+def clone(gitURL, logger, **kwargs):
   """
   Clones the given git repository
 
-  @param gitURL: The repository URL.
-  @param kwargs: Various options to git clone gitURL can be passed as keyword
-    arguments. For now only directory option is handled.
-  e.g.
-  clone(gitURL, directory=nameOfDirectory)
+  :param gitURL: The repository URL.
 
-  @raises
-    infrastructure.utilities.exceptions.CommandFailedError: if
-    the command fails
+  :param logger: logger for additional debug info if desired
 
-  @returns: The blob output of git clone
-  @rtype: string
+  :param directory:
+    Optional. If passed, name of the directory where repository
+    will be cloned.
+
+  :raises infrastructure.utilities.exceptions.CommandFailedError:
+    if the command fails
+
+  :returns: The blob output of git clone
+
+  :rtype: str
   """
-  command = "git clone %s" % gitURL
-  if checkIfOptionSet("directory", **kwargs):
-    command += " " + kwargs["directory"]
-  return executeCommand(command)
+  validKwargs = {"directory"}
+  checkKwargs(kwargs=kwargs, validKwargs=validKwargs)
+  command = ["git", "clone", gitURL]
+  if checkIfOptionSet("directory", kwargs):
+    command.append(kwargs["directory"])
+  return executeCommand(command=command, logger=logger)
 
 
 
-def checkout(pathspec, **kwargs):
+def checkout(pathspec, logger, **kwargs):
   """
   Switches to a given commit-ish
 
-  @param pathspec: The name of the branch (commit-ish)
+  :param pathspec: The name of the branch (commit-ish)
 
-  @param kwargs:
+  :param logger: logger for additional debug info if desired
 
-  @raises
-    infrastructure.utilities.exceptions.CommandFailedError: if
-    the command fails
+  :param new: Boolean. Defaults to False. If True, create a new branch.
+
+  :param orphan:
+    Boolean. Defaults to False. If True, create a new orphan branch.
+    If 'new' is defined, this parameter will be ignored.
```

@shantanoo: if an invalid combination of parameters is specified, then this function should raise TypeError. We don't want to silently suppress invalid args and invalid combinations of args, which are most likely a bug in the caller.